cobc/parser.y

```diff
@@ -12900,15 +12900,21 @@ close_body:
 close_files:
   file_name _close_option
   {
-#if 0 /* CHECKME: likely not needed */
+	/* We need to create a list with a CLOSE statement for every file
+	   within the current_statement instead of nesting them, which
+	   is what would happen if we don't save the current statement
+	   and restore it. */
+	struct cb_statement * saved_current_statement = current_statement ;
 	begin_implicit_statement ();
-#endif
 	cb_emit_close ($1, $2);
+	current_statement = saved_current_statement ;
   }
 | close_files file_name _close_option
   {
+	struct cb_statement * saved_current_statement = current_statement ;
```

Comment:
Where do we need the reset of the current_statement?

Reply:
OK, I see it now - the issue is that `begin_implicit_statement()` both nests the new statement to the `current_statement` and then overwrites that. The approach you've used is reasonable, while not ideal.

Please adjust the signature of the implicit statement as follows:
`begin_implicit_statement (cb_tree real_statement, cb_tree pos)`

The callers then either pass the `saved_current_statement` into it (for lists like in `open_file_entry`) or restore the `current_statement` (like in `close_files`, where we don't have a list).
They also pass "whatever reference creates the implied statement" (in the case above: `file_name`), so the implicit statement of `file-2` and `file-3` in

```cobol
CLOSE file-1
     file-2
     file-3
```

ill point to the correct source file, line and column.

Reply:
I see what you mean for the location `cb_tree pos`, but not for `real_statement`. `begin_implicit_statement` needs to modify `current_statement` at the end, so that the next function will work on it (`cb_emit_open`, `cb_emit_close`, etc.), so it is not possible to restore `current_statement` inside `begin_implicit_statement` (unless we get rid of `current_statement` and pass the statement to all the functions `cb_emit*`, but that would be a big refactoring).

Reply:
`begin_implicit_statement` will use the "original" statement only for attaching the implicit one:

```diff
-	current_statement->body = cb_list_add (current_statement->body,
+	real_statement->body = cb_list_add (real_statement->body,
 					    CB_TREE (new_statement));
```

This should work fine, no?

Reply:
The problem I see is not there, it's the fact that `begin_implicit_statement` needs to modify `current_statement` (because `cb_emit*` functions work on `current_statement`). For example, if I modify the code with the new prototype:

```
| close_files file_name _close_option
  {
	begin_implicit_statement (current_statement,$2);
	cb_emit_close ($2, $3);
  }
```
would not work, because for the next filename, `current_statement` is not the toplevel `CLOSE` anymore, but the one generated by `begin_implicit_statement`, that will be passed to the new call of `begin_implicit_statement`, so that the new statement will be nested inside the previous one.

Reply:
I'm not sure if you like the result (I find it a bit more "clean").